API: Add .mT attribute for arrays

Kai-Striega opened the pull request:
<!--         ----------------------------------------------------------------
                MAKE SURE YOUR PR GETS THE ATTENTION IT DESERVES!
                ----------------------------------------------------------------

*  FORMAT IT RIGHT:
      https://www.numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message

*  IF IT'S A NEW FEATURE OR API CHANGE, TEST THE WATERS:
      https://www.numpy.org/devdocs/dev/development_workflow.html#get-the-mailing-list-s-opinion

*  HIT ALL THE GUIDELINES:
      https://numpy.org/devdocs/dev/index.html#guidelines

*  WHAT TO DO IF WE HAVEN'T GOTTEN BACK TO YOU:
      https://www.numpy.org/devdocs/dev/development_workflow.html#getting-your-pr-reviewed
-->
This is the initial commit for adding an `mT`, or matrix transpose, attribute for arrays. It is still a work in progress and not done yet. Please note that this is my first time working on NumPy's C-API. The PR is mainly being submitted this early to confirm that I am on the right track.

Things that are still needed:

- [x] Release note
- [x] Masked array support
- ~[ ] `np.transpose`~
- [x] Check input dimensions in `array_matrix_transpose_get` and exit early if 1d array
- [x] Tests fail for higher dimensional arrays
- [x] Figure out the correct location for the tests

Another note; `ndarray.T` has a corresponding function `np.transpose`, as part of this PR should we add a corresponding function for `.mT` e.g. `np.matrix_transpose`? If this is the case, should I add a function to `PyArray_MatrixTranspose` to `shape.c` rather than defining the function inside `getset.c` as is done with the regular transpose function?

Kai-Striega committed:
ENH: Add .mT attribute for arrays

Kai-Striega commented on numpy/core/tests/test_matrix_transpose.py:
```diff
@@ -0,0 +1,26 @@
+def test_matrix_transpose_equals_swapaxes(shape):
+    num_of_axes = len(shape)
+    total_elements = np.prod(shape)
+    arr = np.arange(total_elements).reshape(shape)
```
This is pretty naive. It could create a very large array, that may cause problems depending on which shapes are generated.

seberg replied:
Use `broadcast_to`?  Trusts the content is correct, but that seems OK (at least beyond 1-2 explicit tests).  TBH, hypothesis seems a bit overkill in this case, its not like there is any weird complexity we need to cover beyond high dimensions.

Kai-Striega replied:
Done.

Kai-Striega commented on numpy/core/tests/test_matrix_transpose.py:
```diff
@@ -0,0 +1,26 @@
+from numpy.testing import assert_array_equal
+
+
+def test_matrix_transpose_equals_swapaxes_1d():
```
Here and below the test name needs to be renamed to use transpose, not swapaxes.

Kai-Striega committed:
TST: Fix test names to reflect test behaviour

The lower dimensional (1d and 2d) tests compare the behaviour of `.mT` against
`.T` not swapaxes.

Kai-Striega committed:
TST: Increase max_dims for matrix_transpose test

The test generates random shapes of a given length. This was previously used
with min_dims=3. This excludes arrays of dimension 0, 1, 2 which we want.
Instead we test all dimensions up to 7.

seberg commented on numpy/core/src/multiarray/getset.c:
```diff
@@ -931,6 +931,13 @@ array_transpose_get(PyArrayObject *self, void *NPY_UNUSED(ignored))
+array_matrix_transpose_get(PyArrayObject *self, void *NPY_UNUSED(ignored))
+{
+    int ndim = PyArray_NDIM(self);
+    return PyArray_SwapAxes(self, ndim - 2, ndim - 1);
```
Even if it raises, I expect we should have an explicit error anyway for clarity.

Kai-Striega replied:
I'm not sure if an explicit error is the way to go. Without any changes, `.mT` returns the same as `.T` which is what we want, right? I was thinking along the lines of making it explicit (if `ndim == 1` return `self`).

seberg replied:
no, it has to error `ndim < 2`, otherwise it gives the wrong idea in formulas.  Maybe adding a dimension might work, but that would require a lot of convincing, so I think it's a no-go right now, to be clear, I am not convinced yet.  (And I don't actually think it is that clear whether someone wants a column or row vector even if we have an implicit decision anyway.)

Kai-Striega replied:
Ok. I'll follow your advice. Would a `ValueError` be a good fit here? I've had a quick look through the available errors and couldn't find anything that fits better.

seberg replied:
I think that is what we use for shape related errors always, so that sounds good to me.  We were discussing introducing a shape or broadcast error, but that is a different project.

seberg commented on numpy/core/tests/test_matrix_transpose.py:
```diff
@@ -0,0 +1,26 @@
+import pytest
```
There is probably no great place... Could start an `test_arrayobject.py` if you don't find a nice place elsewhere.

Kai-Striega replied:
Done.

seberg reviewed (commented):
Looks like a good start, will mark as draft for now if you don't mind, just unmark it when things are settling (not that I think there is much missing)!

Kai-Striega commented on numpy/core/tests/test_matrix_transpose.py:
```diff
@@ -0,0 +1,26 @@
+    total_elements = np.prod(shape)
+    arr = np.arange(total_elements).reshape(shape)
+    tgt = np.swapaxes(arr, num_of_axes - 2, num_of_axes - 1)
+    mT = arr.mT
```
This test now passes. Turns out CLion doesn't rebuild every time I run the test suite. I was using an outdated build.

Kai-Striega committed:
TST: Use `np.broadcast_to` to save memory in test

This commit reduces the memory required by the test. Instead of explicitly
constructing an array of shape `shape`, we now construct only the last dim
and broadcast over that.

Kai-Striega committed:
TST: Avoid using hypothesis in test case

Hypothesis seems to be overkill for the provided situation. Use explicitly
defined values instead.

Kai-Striega committed:
TST/MAINT: Rename file to be more generic

Kai-Striega committed:
ENH: Raise ValueError for ndim < 2 in .mT

This commit causes `.mT` to raise a ValueError if ndim < 2. This is because
the matrix transpose is undefined for 0 or 1 dimensional arrays and having
it implemented may lead to confusion by users.

rgommers commented:
Thanks for working on this Kai!

> Another note; `ndarray.T` has a corresponding function `np.transpose`, as part of this PR should we add a corresponding function for `.mT` e.g. `np.matrix_transpose`?

I'd say yes we want that (because of the analogue to `.T` and because [the array API standard](https://data-apis.org/array-api/latest/API_specification/generated/array_api.matrix_transpose.html) has it). It doesn't have to be in this PR though, if it's already large and easier to do as a follow-up - whatever seems best to you.

Kai-Striega committed:
TST: Add test for the existence of `matrix_transpose`

Kai-Striega committed:
ENH: Try to create a `matrix_transpose` function

Kai-Striega commented on numpy/core/code_generators/numpy_api.py:
```diff
@@ -370,6 +370,8 @@ def get_annotations():
     'PyDataMem_SetHandler':                 (304, MinVersion("1.22")),
     'PyDataMem_GetHandler':                 (305, MinVersion("1.22")),
     # End 1.22 API
+    'PyArray_MatrixTranspose':              (306, MinVersion("1.25")),  # TODO: What does `MinVersion` do?
```
I've done some reading on what `MinVersion` does. It's defined in numpy/core/code_generators/genapi.py

seberg replied:
That is correct, but I am not sure I feel we need to add this?  At some point people aimed to have a full featured C-API, I at least don't really feel there is a point for many convenience functions in C.

seberg replied:
MinVersion says that this is the first version that defines the function.

Kai-Striega replied:
> but I am not sure I feel we need to add this? 

I've decided it's not worth adding. My idea was to copy `np.transpose` so that I have something to go off. That isn't the way the reviewers want me to go, I'll leave it out.

> MinVersion says that this is the first version that defines the function.

Thanks

Kai-Striega commented on numpy/core/src/multiarray/methods.c:
```diff
@@ -2399,6 +2399,12 @@ array_transpose(PyArrayObject *self, PyObject *args)
+static PyObject *
+array_matrix_transpose(PyArrayObject *self)
+{
+    return PyArray_MatrixTranspose(self);
```
Does this need to be explicitly cast to `PyObject *`

seberg replied:
No, since `PyArray_MatrixTranspose` returns `PyObject *`, so it matches.  Otherwise you would see a compiler warning anyway.

Kai-Striega commented:
@rgommers I'm happy to work on it in this PR, but I'm still learning about NumPy internals, meaning that the commit history could become quite messy as I learn (of course I'll try my best to keep it clean). As a summary, I think we should match the functions for `transpose` i.e.:

* `np.matrix_transpose`
* `np.ndarray.mT`
* `np.ndaray.matrix_transpose`
* `np.ma.matrix_transpose`
* `np.ma.MaskedArray.mT`
* `np.ma.MaskedArray.matrix_transpose`

The next step for me is to add `np.ndaray.matrix_transpose`, which I am having some trouble with. What I have done so far is:

* Moved the implementation into a new function (`PyArray_MatrixTranspose`) into `numpy/core/src/multiarray/shape.c` alongside `PyArray_Transpose` 
* Added a `matrix_transpose` entry to `array_methods` in `numpy/core/src/multiarray/methods.c`
* Added `PyArray_Transpose` to the `multiarray_funcs_api` in `numpy/core/code_generators/numpy_api.py`

I'm not really sure if this is the right direction, or if we even want to add `PyArray_MatrixTranspose` to the C-API as (I think) I am doing.

mattip commented:
I would not touch the NumPy C-API with this PR. It is enough to add the attribute to ndarray and ma. If in the future we wish to export it via the C-API, that can be another PR.

rgommers commented:
> As a summary, I think we should match the functions for `transpose` i.e.:

We don't need/want both a `.mT` property and a `.matrix_transpose` method. It's duplicate and the plan is to trim down on `ndarray` methods. Just the property and the `np.matrix_transpose` function (and their MaskedArray equivalents) is enough.

charris commented:
I rather like the idea of an `.mT` property :) It would be useful to have a short way to call it in `...@...@...` constructions.

Kai-Striega committed:
REV: Remove PyArray_MatrixTranspose from C-API

This was decided to not be needed in the PR review.
Skipping CI as I know it doesn't work and don't need
CI to tell me that.

[skip ci]

Kai-Striega commented:
I've removed the code adding it to the C-API (I think) and stopped working on adding an `ndarray.matrix_transpose` method. I've kept the function `PyArray_MatrixTranspose` in `shape.c` as I want to be able to share it between `ndarray.mT` and `np.matrix_transpose` - this is causing the test to segfault compared to the earlier version. 

I may need some help structuring the code so that I can share it between `ndarray.mT` and `np.matrix_transpose`.

Kai-Striega committed:
DOC: Add release note for matrix transpose

[skip ci]

seberg commented:
You have to add the new function to a header file now and include it, probably.  If you check the test failures, you see they fail on compiler warnings.

BvB93 reviewed (commented):
Can you add `ndarray.mT` annotations to the main __init__.pyi stub file? The type is identical to `ndarray.T` so you can just reuse that it:
https://github.com/numpy/numpy/blob/cc0abd768575d7f9e862de0b4912af27f6e9690d/numpy/__init__.pyi#L986-L987

Kai-Striega committed:
TYP: Add typehint for new `.mT` attribute

Kai-Striega committed:
ENH: Include PyArray_MatrixTranspose in shape.h

Kai-Striega committed:
Merge branch 'numpy:main' into mT

Kai-Striega committed:
ENH/TST: First attempt to support MaskedArrays

Kai-Striega commented:
@seberg I've *finally* got the masked array implementation working! I'm not sure if it's done correctly, could you take a look at my last commit? 

If it's right, I think I have three things left:

- [ ] Take the reference to the function versions out of the release notes
- [ ] Add the documentation
- [ ] Add type hints for the masked array

seberg commented on numpy/ma/core.py:
```diff
@@ -6094,6 +6094,22 @@ def take(self, indices, axis=None, out=None, mode='raise'):
+
+        if self._mask is nomask:
+            # TODO: Should this return a masked_array or regular array?
+            return self._data.mT
```
A masked array.  Otherwise, seems good.  Although `.transpose()` has some `_update_from_self` stuff, it may make sense to just stick close to it in the off-chance it matters.

Kai-Striega replied:
I'm struggling to replicate `.transpose()` as it uses the function version, not the attribute. I haven't been figured out how to do this (yet) and will leave it for another PR - maybe I can come back and tidy it up then if there are problems?

Kai-Striega committed:
ENH: Make ma.MaskedArray.mT return a masked array in all cases

Kai-Striega committed:
TYP: ma: Add typehint for .mT

Kai-Striega committed:
DOC: Document mT property for masked arrays

Kai-Striega committed:
REL: Remove matrix_transpose function from release note

This function will be added in another PR.

Kai-Striega committed:
DOC: Add documentation for `ndarray.mT`

Kai-Striega commented:
@rgommers @seberg I've made the last changes that (I think) are needed, could you please ask you two for another review?

I've decided to leave the `np.matrix_transpose` function for now, I'll get to it in another PR

seberg commented on numpy/core/_add_newdocs.py:
```diff
@@ -2927,6 +2927,43 @@
+    
+    The matrix transpose is the transpose of the last two dimensions, even
+    if the array is of higher dimension.
+
```
```suggestion

    .. versionadded:: 2.0

```

seberg commented on numpy/ma/core.py:
```diff
@@ -6094,6 +6094,39 @@ def take(self, indices, axis=None, out=None, mode='raise'):
+
+        The matrix transpose is the transpose of the last two dimensions, even
+        if the array is of higher dimension.
+
```
```suggestion

    .. versionadded:: 2.0

```

seberg reviewed (approved):
LGTM, thanks, just should add the versionadded.

Kai-Striega committed:
DOC: Document version added

Co-authored-by: Sebastian Berg <sebastian@sipsolutions.net>

seberg commented:
Thanks, lets put it in.

After merging, I noticed that the docs actually render slightly incorrectly.  Not a big deal, but maybe you want to follow up?

I guess we need `<BLANKLINE>` or maybe its fine to just remove the blank line also.

Kai-Striega commented:
Thanks that's my first C level contribution 👌. Let's hope for many more.

I'll follow up with the docs rendering properly.
